src/main/java/co/cask/http/RequestRouter.java

```diff
@@ -16,132 +16,91 @@
-    }
-  }
-
-  private class SSLHandshakeListener implements ChannelFutureListener {
-    @Override
-    public void operationComplete(ChannelFuture channelFuture) throws Exception {
-      if (!channelFuture.isSuccess()) {
-        channelFuture.addListener(ChannelFutureListener.CLOSE);
+    if (!(msg instanceof HttpRequest)) {
+      // If there is no methodInfo, it means the request was already rejected.
+      // What we received here is just residue of the request, which can be ignored.
+      if (methodInfo != null) {
+        super.channelRead(ctx, msg);
       }
+      return;
     }
-  }
-  /**
-   * If return type of the handler http method is BodyConsumer, remove "HttpChunkAggregator" class if it exits
-   * Else add the HttpChunkAggregator to the pipeline if its not present already.
-   */
-  private boolean handleRequest(HttpRequest httpRequest, Channel channel, ChannelHandlerContext ctx) throws Exception {
-
-    methodInfo = httpMethodHandler.getDestinationMethod(
-      httpRequest, new BasicHttpResponder(channel, HttpHeaders.isKeepAlive(httpRequest)));
+    HttpRequest request = (HttpRequest) msg;
+    BasicHttpResponder responder = new BasicHttpResponder(ctx.channel(), sslEnabled);
+    methodInfo = prepareHandleMethod(request, responder, ctx);
 
-    if (methodInfo == null) {
-      return false;
-    }
-    ctx.setAttachment(methodInfo);
-    if (methodInfo.isStreaming()) {
-      if (channel.getPipeline().get("aggregator") != null) {
-        channel.getPipeline().remove("aggregator");
-      }
+    if (methodInfo != null) {
+      ctx.fireChannelRead(msg);
     } else {
-      if (channel.getPipeline().get("aggregator") == null) {
-        channel.getPipeline().addAfter("router", "aggregator", new HttpChunkAggregator(chunkMemoryLimit));
+      if (!responder.isResponded()) {
+        // If not yet responded, just respond with a not found and close the connection
+        HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_FOUND);
+        ctx.channel().writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
+      } else {
+        // If already responded, just close the connection
+        ctx.channel().close();
       }
     }
-    return true;
   }
 
-
-  public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) {
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
     String exceptionMessage = "Exception caught in channel processing.";
-    Throwable cause = e.getCause();
     if (exceptionRaised.compareAndSet(false, true)) {
       if (methodInfo != null) {
         LOG.error(exceptionMessage, cause);
         methodInfo.sendError(HttpResponseStatus.INTERNAL_SERVER_ERROR, cause);
         methodInfo = null;
       } else {
-        ChannelFuture future = Channels.future(ctx.getChannel());
-        future.addListener(ChannelFutureListener.CLOSE);
```

Comment:
why move the `addListener` call to the two cases below?

Reply:
We don't need to create a new future. The `ctx.writeAndFlush` already returns a future for that operation.